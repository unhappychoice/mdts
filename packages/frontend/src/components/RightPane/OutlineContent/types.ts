export interface OutlineItem {
  id: string;
  level: number;
  content: string;
}
