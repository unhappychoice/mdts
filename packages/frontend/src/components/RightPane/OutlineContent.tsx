import { CircularProgress, List, ListItem, ListItemText, Box } from '@mui/material';
import React from 'react';

interface OutlineItem {
  id: string;
  level: number;
  content: string;
}

interface OutlineContentProps {
  outline: OutlineItem[];
  loading: boolean;
  error: string | null;
  onItemClick: (id: string) => void;
}

const OutlineContent: React.FC<OutlineContentProps> = ({ outline, loading, error, onItemClick }) => {
  if (loading) {
    return (
      <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: 'calc(100vh - 132px)' }}>
        <CircularProgress />
      </Box>
    );
  }

  if (error) {
    return <p>Error: {error}</p>;
  }

  return (
    <List dense sx={{ overflowY: 'auto', height: 'calc(100vh - 132px)' }}>
      {outline.map((item, index) => (
        <ListItem button key={`${item.id}-${index}`} sx={{ pl: item.level * 2 }} onClick={() => onItemClick(item.id)}>
          <ListItemText
            primary={item.content}
            sx={{
              overflow: 'hidden',
              whiteSpace: 'nowrap',
              textOverflow: 'ellipsis',
            }}
          />
        </ListItem>
      ))}
    </List>
  );
};

export default OutlineContent;
