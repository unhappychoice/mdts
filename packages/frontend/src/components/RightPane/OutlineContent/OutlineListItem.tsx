import { ListItem, ListItemText } from '@mui/material';
import React, { useCallback } from 'react';
import { OutlineItem } from './types';

interface OutlineListItemProps {
  item: OutlineItem;
  onItemClick: (id: string) => void;
}

export const OutlineListItem: React.FC<OutlineListItemProps> = ({ item, onItemClick }) => {
  const handleClick = useCallback(() => {
    onItemClick(item.id);
  }, [onItemClick, item.id]);

  return (
    <ListItem button key={item.id} sx={{ pl: item.level * 2 }} onClick={handleClick}>
      <ListItemText primary={item.content} sx={{ overflow: 'hidden', whiteSpace: 'nowrap', textOverflow: 'ellipsis' }} />
    </ListItem>
  );
};
