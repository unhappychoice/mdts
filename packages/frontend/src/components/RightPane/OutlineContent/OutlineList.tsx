import { List } from '@mui/material';
import React from 'react';
import { OutlineListItem } from './OutlineListItem';
import { OutlineItem } from './types';

interface OutlineListProps {
  outline: OutlineItem[];
  onItemClick: (id: string) => void;
}

export const OutlineList: React.FC<OutlineListProps> = ({ outline, onItemClick }) => (
  <List dense className="custom-scrollbar" sx={{ overflowY: 'auto', height: 'calc(100vh - 132px)', px: 2 }}>
    {outline.map((item) => (
      <OutlineListItem key={item.id} item={item} onItemClick={onItemClick} />
    ))}
  </List>
);
