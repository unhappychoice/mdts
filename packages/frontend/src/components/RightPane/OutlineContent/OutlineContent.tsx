import React from 'react';
import ErrorView from '../../ErrorView';
import { LoadingIndicator } from './LoadingIndicator';
import { OutlineList } from './OutlineList';
import { OutlineItem } from './types';

interface OutlineContentProps {
  outline: OutlineItem[];
  loading: boolean;
  error: string | null;
  onItemClick: (id: string) => void;
}

const OutlineContent: React.FC<OutlineContentProps> = ({ outline, loading, error, onItemClick }) => {
  if (loading) {
    return <LoadingIndicator />;
  }

  if (error) {
    return <ErrorView error={error} />;
  }

  return <OutlineList outline={outline} onItemClick={onItemClick} />;
};

export default OutlineContent;
