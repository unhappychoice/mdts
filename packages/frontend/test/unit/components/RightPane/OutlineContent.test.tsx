import { render, screen, fireEvent } from '@testing-library/react';
import React from 'react';
import OutlineContent from '../../../../src/components/RightPane/OutlineContent';

describe('OutlineContent', () => {
  const mockOutline = [
    { id: 'heading-1', content: 'Heading 1', level: 1 },
    { id: 'subheading-1', content: 'Subheading 1', level: 2 },
    { id: 'heading-2', content: 'Heading 2', level: 1 },
  ];

  test('renders loading spinner when loading is true', () => {
    render(
      <OutlineContent
        outline={[]}
        loading={true}
        error={null}
        onItemClick={jest.fn()}
      />
    );
    expect(screen.getByRole('progressbar')).toBeInTheDocument();
  });

  test('renders error message when error is present', () => {
    render(
      <OutlineContent
        outline={[]}
        loading={false}
        error="Failed to load outline"
        onItemClick={jest.fn()}
      />
    );
    expect(screen.getByText('Error: Failed to load outline')).toBeInTheDocument();
  });

  test('renders outline items correctly', () => {
    const { asFragment } = render(
      <OutlineContent
        outline={mockOutline}
        loading={false}
        error={null}
        onItemClick={jest.fn()}
      />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('Heading 1')).toBeInTheDocument();
    expect(screen.getByText('Subheading 1')).toBeInTheDocument();
    expect(screen.getByText('Heading 2')).toBeInTheDocument();
  });

  test('calls onItemClick when an outline item is clicked', () => {
    const onItemClickMock = jest.fn();
    render(
      <OutlineContent
        outline={mockOutline}
        loading={false}
        error={null}
        onItemClick={onItemClickMock}
      />
    );
    fireEvent.click(screen.getByText('Heading 1'));
    expect(onItemClickMock).toHaveBeenCalledWith('heading-1');
  });

  test('applies correct padding based on item level', () => {
    render(
      <OutlineContent
        outline={mockOutline}
        loading={false}
        error={null}
        onItemClick={jest.fn()}
      />
    );
    const subheadingItem = screen.getByText('Subheading 1').closest('li');
    expect(subheadingItem).toHaveStyle('padding-left: 32px'); // level 2 * 16px (default ListItem padding) = 32px
  });
});
