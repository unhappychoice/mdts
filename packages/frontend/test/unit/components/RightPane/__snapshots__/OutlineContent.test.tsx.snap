// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`OutlineContent renders outline items correctly 1`] = `
<DocumentFragment>
  <ul
    class="MuiList-root MuiList-padding MuiList-dense css-h9oyac-MuiList-root"
  >
    <li
      class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding css-1s7t0kb-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root MuiListItemText-dense css-6qv7tz-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-2wbpcv-MuiTypography-root"
        >
          Heading 1
        </span>
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding css-1bz6dlg-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root MuiListItemText-dense css-6qv7tz-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-2wbpcv-MuiTypography-root"
        >
          Subheading 1
        </span>
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding css-1s7t0kb-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root MuiListItemText-dense css-6qv7tz-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-2wbpcv-MuiTypography-root"
        >
          Heading 2
        </span>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
